### ISSUE TITLE
Error: Database error: Foreign key constraint failed

### ISSUE THREAD

**Issue by JaapWeijland** [prisma/prisma#6995]
### Bug description

I have the following schema:

```
model User {
  id                String @id @default(uuid())
  ...
  address           Address? @relation(fields: [id], references:[entityId])
}

model Address {
  entityId    String @id
  city        String?
  line1       String?
  line2       String?
  postalCode  String?
  user        User?
}
```

I execute: `npx prisma db push`

I am using: Postgres

The error I get:

```
Error: Database error
Foreign key constraint failed: User_id_fkey
   0: sql_migration_connector::sql_database_step_applier::apply_migration
             at migration-engine/connectors/sql-migration-connector/src/sql_database_step_applier.rs:15
   1: migration_core::api::SchemaPush
             at migration-engine/core/src/api.rs:163
```

Am I missing something conceptually? The data model seems valid to me. I use the VS Code plugin to lint my schema and it's not showing any errors.

### How to reproduce

As seen above.

### Expected behavior

The new data model to be pushed correctly to the database.

### Prisma information

```
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}
```

Version: `"prisma": "^2.22.1",`

### Environment & setup

- OS: MacOS
- Database: PostgreSQL
- Node.js version: 12.14.0
- Prisma version: 2.22.1

```
prisma               : 2.22.1
@prisma/client       : 2.21.2
Current platform     : darwin
Query Engine         : query-engine 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/query-engine-darwin)
Migration Engine     : migration-engine-cli 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine : introspection-core 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary        : prisma-fmt 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash : 60cc71d884972ab4e897f0277c4b84383dddaf6c
Studio               : 0.379.0
```

**Comment by JaapWeijland** [comment]
Update: I found out that if I'd replace the `@relation` tag, it works:

```
model User {
  id                String @id @default(uuid())
  ...
  address           Address?
}

model Address {
  entityId    String @id
  city        String?
  line1       String?
  line2       String?
  postalCode  String?
  user        User? @relation(fields: [entityId], references:[id])
}
```

Is this implemented deliberately? If so, the error returned should be a bit more explanatory.

**Comment by janpio** [comment]
This is interesting. In your initial schema you used the primary key of the table/model as the foreign key pointing to the primary key of the other table - and that indeed looks valid to me. We will have to reproduce this and understand deeper what is going wrong.

**Comment by pantharshit00** [comment]
This is not reproducible in an empty database, did you have any data when you tried this initially?

**Comment by JaapWeijland** [comment]
@pantharshit00 Yes I had. Is there anything I can report to help finding the bug?

**Comment by pantharshit00** [comment]
If by any chance you can share a SQL dump of the DB if it doesn't contain any sensitive information, it would be very helpful. Otherwise I will try with some dummy data.

**Comment by rizkyramadhan** [comment]
I think I stumble for this problem too

**Comment by rizkyramadhan** [comment]
I will send schema.prisma to privately.

**Comment by janpio** [comment]
Did that happen @rizkyramadhan?

**Comment by rizkyramadhan** [comment]
I think I figured out the cause of my problem. Somehow prisma db pull is wrong when guessing primary key type. My PK is in Integer, but the generated code is using decimal(10,0). Manually changing that fixes this problem.

**Comment by janpio** [comment]
Can you share the SQL of that database/table? That would be super interesting for us to reproduce as it really should not happen.

**Comment by rizkyramadhan** [comment]
My database is huge, I think dumping it is not practical. I don't mind sharing the username/password with you - but not in public, as it is on my playground dev server.

**Comment by janpio** [comment]
Only the schema of course - not the data. But feel free to send the connection string to jan@prisma.io - and I can take a look. Please reference this issue in the email. Thanks.

**Comment by janpio** [comment]
Unfortunately not @joshsny. Can you create a new bug issue with your information?  We would love to investigate this and figure out what is going on for your `db pull`. Thanks.

**Comment by archcra** [comment]
I have encountered this error, and googled a lot to find the answer. The problem, as it turned out, is an EBKAC.
I designed the Prisma schema correctly, but wrong with the CREATE TABLE SQL. 
schema: 
model Array_plate_cell {
...
  array_plate Array_plate? @relation(fields: [plate_rid], references: [id])

SQL:
CREATE TABLE "Array_plate_cell" (
 ...
    FOREIGN KEY ("plate_rid") REFERENCES "Plate" ("id") ON DELETE CASCADE ON UPDATE CASCADE,

and the "Plate" table in SQL, should be "Array_plate" table:

This EBKAC was not easy to find out because the two schemas are very similar: 
model Array_plate {
  id                Int                @id @default(autoincrement())
  plate_id          String 
...

model Plate {
  id                Int           @id @default(autoincrement())
  plate_id          String 
...

Hope this information helps.

**Comment by janpio** [comment]
So in your case @archcra it was caused by a mismatch between your SQL schema pointing the Foreign Key to one table, and the Prisma Schema relation pointing to another model?

**Comment by Gasparuss** [comment]
### When I try to delete my account in prisma studio I get such error:
 Foreign key constraint failed on the field: `UserToken_userId_fkey (index)`



<pre>
model User {
  id                 Int                  @id @default(autoincrement())
  name               String
  email              String
  password           String
  isActive           Boolean?             @default(false)
  createdAt          DateTime             @default(dbgenerated("(now())::timestamp without time zone")) @db.Timestamp(6)
  updatedAt          DateTime             @default(dbgenerated("(now())::timestamp without time zone")) @db.Timestamp(6)
  UserMovieFavourite UserMovieFavourite[]
  UserMovieRate      UserMovieRate[]
  UserToken          UserToken[]
}

model UserToken {
  userId Int
  token  String @db.VarChar(255)
  User   User   @relation(fields: [userId], references: [id])
  @@id([userId, token])
}
</pre>

**Comment by Lucaspaula6** [comment]
It happened with me. I was getting the same error when creating objects inside objects in a relation from one to many. 

The issue has been causd by the hierarchaly minor object beeing created before the higher one. 

The solution was to seed manually object by object. 

Hope it helped. 

Obs.: I thought about make an logical sequence that allows the subsquent seed to happen only after previous. Didn't implement anything like that. If somebody implemented this solution or have a better one, please let me know.

Obrigado. =)
lucas_araujo_11@hotmail.com

**Comment by cvolzer3** [comment]
For those of you that have made it here and haven't found a solution, my problem was with how my Prisma query was written. Instead of plainly adding the foreign key's identifier in the data field (A), I changed to `connect` (B).

For some reason, my development environment was allowing me to successfully create an entry with the first implementation. It wasn't until it was deployed that it would error out. Same Prisma version, same database. 🤷🏻‍♂️

**Schema:**
```
model user {
  userId     String    @id @default(cuid())
  hospitalId String
  firstName String
  hospital   hospital? @relation(fields: [hospitalId], references: [hospitalId])
}

model hospital {
  hospitalId String @id @default(cuid())
  users      user[]
}
```

**(A) Inlined foreign identifier:**
Worked in local dev, failed in production
```
  const user = await prisma.user.create({
    data: {
      firstName,
      hospitalId
    }
  });
```

**(B) Connect records:**
Worked in both local dev and in production
```
  const user = await prisma.user.create({
    data: {
      firstName,
      hospital: {
        connect: {
          hospitalId
        }
      }
    }
  });
```

**Comment by archcra** [comment]
您的邮件已被接收，我将努力找时间处理。

**Comment by archcra** [comment]
您的邮件已被接收，我将努力找时间处理。

**Comment by orta** [comment]
To add to this pile, I accidentally re-used the same relation name between two different models which ran fine but fails at runtime:

e.g. this pseudo prisma file:

```
model User {
  posts Post[] @relation("author")
}
model Company {
  posts Post[] @relation("author")
}
```

**Comment by archcra** [comment]
您的邮件已被接收，我将努力找时间处理。

**Comment by janpio** [comment]
Can you quickly split this out into a standalone issue @orta? Seems we are missing a validation here.

**Comment by foureight84** [comment]
I came across this ticket with the same issue after updating my schema from using Int @id to uuid @id

I've  tested this several times, starting with a fresh db. and performing a prisma db ush with the old schema and then new schema. With the old schema where all the primary keys are Int, I don't have an issue with Create. But once I repeat the process with a clean DB but uuid primary key schema then I get this error with Create.

**Comment by ColonelEagel** [comment]
If there is someone still facing this issue, Before proceeding with the migration, you can manually drop the existing foreign key constraint in your database. However, be cautious when doing this, as it can affect data integrity. Run the following SQL command to drop the constraint:

`
ALTER TABLE  <ChildTable>  DROP FOREIGN KEY  <ForeignKeyName`> ;
`

Replace <ChildTable> with the name of the table containing the constraint you wish to remove and <ForeignKeyName> with the actual name of the foreign key constraint that needs to be dropped

**Comment by Leni-Vienne** [comment]
In my case, the problem was that two of my tables, **that had nothing to do with what I was trying to insert**, were dependent of each other ("users are part of an org" and "orgs have a president user") and one of the tables had the relation field mandatory and with a default value, which indeed didn't make sense.

```
model user {
   id_user      Int        @id @default(autoincrement())
   name        String   @db.VarChar(50)
   id_org       Int
   org           org?     @relation(fields: [id_org], references: [id_org], onDelete: NoAction, onUpdate: NoAction)
   president org[]     @relation("bureau_chef") // <---
}

model org {
    id_org         Int       @id @default(autoincrement())
    org name   String  @db.VarChar(100)
    president    Int       @default(1) // <--- the root cause, should have been optional with no @default value
}

```
